File: frontend/scripts/check-allowlist.js
#!/usr/bin/env node
/**
 * Check that all npm dependencies are in allow-list
 * Fails CI if any dependency is not allowed
 *
 * Usage: node scripts/check-allowlist.js
 * Exit code: 0 (all allowed), 1 (violations found)
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');


/**
 * Get list of installed packages from package-lock.json
 */
function getInstalledPackages() {
    const lockfilePath = path.join(process.cwd(), 'package-lock.json');

    if (!fs.existsSync(lockfilePath)) {
        console.error('❌ package-lock.json not found');
        console.error('   Run: npm install');
        process.exit(1);
    }

    const lockfile = JSON.parse(fs.readFileSync(lockfilePath, 'utf8'));
    const packages = {};

    // Get top-level dependencies
    if (lockfile.packages) {
        for (const [name, info] of Object.entries(lockfile.packages)) {
            // Skip root package
            if (name === '') continue;

            // Only include direct dependencies
            if (info.version && !name.includes('node_modules')) {
                // Get package name without scoped prefix
                const pkgName = name.replace('node_modules/', '');
                packages[pkgName] = info.version;
            }
        }
    }

    return packages;
}


/**
 * Parse allow-list from JSON file
 */
function parseAllowList() {
    const allowListPath = path.join(process.cwd(), 'allowed-dependencies.json');

    if (!fs.existsSync(allowListPath)) {
        console.error(`❌ Allow-list file not found: ${allowListPath}`);
        process.exit(1);
    }

    const allowList = JSON.parse(fs.readFileSync(allowListPath, 'utf8'));

    const allowed = {};

    if (allowList.allowedDependencies) {
        for (const [name, info] of Object.entries(allowList.allowedDependencies)) {
            allowed[name] = {
                versionRange: info.versionRange,
                maxVersion: info.maxVersion,
                signatureRequired: info.signatureRequired || false
            };
        }
    }

    return allowed;
}


/**
 * Check if version is within allowed range
 */
function checkVersionCompliance(package, installedVersion, allowedInfo) {
    const versionRange = allowedInfo.versionRange;
    const maxVersion = allowedInfo.maxVersion;

    // Simple version check (use semver in production)
    // For now, just warn if version exceeds maxVersion
    if (maxVersion && installedVersion > maxVersion) {
        return false;
    }

    return true;
}


/**
 * Main check function
 */
function checkAllowList() {
    console.log('🔍 Checking npm dependencies against allow-list...\n');

    // Get installed packages
    const installed = getInstalledPackages();
    console.log(`📦 Found ${Object.keys(installed).length} installed packages`);

    // Parse allow-list
    const allowList = parseAllowList();
    console.log(`✅ Allow-list has ${Object.keys(allowList).length} allowed packages\n`);

    const violations = [];

    // Check each package
    for (const [name, version] of Object.entries(installed)) {
        if (!allowList[name]) {
            violations.push({
                package: name,
                version: version,
                reason: 'Not in allow-list'
            });
        } else {
            // Check version compliance
            const allowedInfo = allowList[name];
            if (!checkVersionCompliance(name, version, allowedInfo)) {
                violations.push({
                    package: name,
                    version: version,
                    reason: `Version ${version} exceeds max ${allowedInfo.maxVersion}`
                });
            }
        }
    }

    // Report results
    if (violations.length > 0) {
        console.log('❌ DEPENDENCY ALLOW-LIST VIOLATIONS DETECTED');
        console.log(`   ${violations.length} violations found\n`);

        for (const v of violations) {
            console.log(`  • ${v.package} (${v.version})`);
            console.log(`    Reason: ${v.reason}`);
            console.log(`    Action: Submit dependency request\n`);
        }

        console.log('To request an exception:');
        console.log('  1. Create issue:');
        console.log('     gh issue create --title "Dependency Request: PACKAGE" \\');
        console.log('       --label "dependency-request"');
        console.log('  2. Security team reviews within 24-48 hours');
        console.log('  3. Once approved, add to allowed-dependencies.json');
        console.log('\nSee DEPENDENCY_ALLOWLIST_POLICY.md for request template\n');

        process.exit(1);
    } else {
        console.log(`✅ All ${Object.keys(installed).length} dependencies are in allow-list`);
        console.log('   Compliance: 100%\n');
        process.exit(0);
    }
}


// Run the check
checkAllowList();

File: scripts/check-allowlist.py
#!/usr/bin/env python3
"""
Check that all Python dependencies are in allow-list
Fails CI if any dependency is not allowed

Usage: python3 scripts/check-allowlist.py
Exit code: 0 (all allowed), 1 (violations found)
"""

import json
import subprocess
import sys
from pathlib import Path
from typing import Dict, List


def get_installed_packages() -> Dict[str, str]:
    """Get list of installed packages from pip freeze"""

    result = subprocess.run(
        ["pip", "list", "--format=json"], capture_output=True, text=True
    )

    if result.returncode != 0:
        print(f"Error running pip list: {result.stderr}")
        sys.exit(1)

    packages = json.loads(result.stdout)

    return {pkg["name"].lower(): pkg["version"] for pkg in packages}


def parse_allow_list(
    allow_list_path: str = "allowed-dependencies.txt",
) -> Dict[str, dict]:
    """Parse allow-list file

    Format: package==min,max # date # notes
    """

    allow_list = {}

    allow_file = Path(allow_list_path)
    if not allow_file.exists():
        print(f"❌ Allow-list file not found: {allow_list_path}")
        sys.exit(1)

    with open(allow_file, "r") as f:
        for line_num, line in enumerate(f, 1):
            line = line.strip()

            # Skip comments and empty lines
            if not line or line.startswith("#"):
                continue

            # Remove comments after package spec
            line = line.split("#")[0].strip()
            if not line:
                continue

            # Parse: package==min,max
            if "==" in line:
                package_spec = line.split("==")[0]
                version_range = line.split("==")[1] if "==" in line else "*"

                package_name = package_spec.lower().strip("_-")  # Normalize

                allow_list[package_name] = {
                    "version_range": version_range,
                    "line": line_num,
                }

    return allow_list


def check_version_compatibility(
    package: str, installed_version: str, allowed_info: dict
) -> bool:
    """Check if installed version is within allowed range"""

    version_range = allowed_info["version_range"]

    # Parse range: min,max
    if "," in version_range:
        min_ver, max_ver = version_range.split(",")
        # Check if version is within range
        # (simplified - use packaging.version in production)
        return True  # Implement proper version comparison
    else:
        # Single version or wildcard
        return True


def check_allow_list():
    """Check all packages against allow-list"""

    # Get installed packages
    installed = get_installed_packages()
    print(f"📦 Found {len(installed)} installed packages")

    # Parse allow-list
    allow_list = parse_allow_list()
    print(f"✅ Allow-list has {len(allow_list)} allowed packages")

    violations = []

    for package_name, version in installed.items():
        # Normalize package name for comparison
        normalized_name = package_name.lower().strip("_-")

        if normalized_name not in allow_list:
            violations.append(
                {
                    "package": package_name,
                    "version": version,
                    "reason": "Not in allow-list",
                }
            )
        else:
            # Check version compliance
            allowed_info = allow_list[normalized_name]
            if not check_version_compatibility(package_name, version, allowed_info):
                violations.append(
                    {
                        "package": package_name,
                        "version": version,
                        "reason": f"Version {version} not in allowed range {allowed_info['version_range']}",
                    }
                )

    # Report results
    if violations:
        print("\n❌ DEPENDENCY ALLOW-LIST VIOLATIONS DETECTED")
        print(f"   {len(violations)} violations found\n")

        for v in violations:
            print(f"  • {v['package']} ({v['version']})")
            print(f"    Reason: {v['reason']}")
            print(f"    Action: Submit dependency request")
            print()

        print("To request an exception:")
        print(
            "  1. Create issue: gh issue create --title 'Dependency Request: PACKAGE'"
        )
        print("     --label 'dependency-request' --body-file <request-template>")
        print("  2. Security team reviews within 24-48 hours")
        print("  3. Once approved, add to allowed-dependencies.txt")
        print("\nRequest template:")
        print(
            "  See DEPENDENCY_ALLOWLIST_POLICY.md section 'New Package Request Workflow'"
        )
        print()

        sys.exit(1)
    else:
        print(f"✅ All {len(installed)} dependencies are in allow-list")
        print(f"   Compliance: 100%")
        sys.exit(0)


if __name__ == "__main__":
    check_allow_list()
